### ISSUE TITLE
ZIO Test: Add Laws

### ISSUE THREAD

**Issue by jdegoes** [zio/zio#3035]
Why can't laws be values? /cc @adamgfraser 

```scala
trait Laws[-R, -A] { self =>
  def run(gen: Gen[A]): ZIO[R, Nothing, TestResult] 

  def + [R1 <: R, A1 <: A](that: Laws[R1, A1]): Laws[R1, A1] = ???
}
object Laws {
  ...
  def apply[A](label: String)(f: (A, A, A) => TestResult): Laws[Any, A] = ???
  ...
}
...
val transitivityLaw = 
  Laws("transitivity law")((a1, a2, a3) => (a1 === a2) && (a2 === a3) ==> (a1 === a3))
...
val allLaws = transitivityLaw + symmetryLaw + ...
```

Maybe with an added `zio.test` operator:

```scala
checkLaws(allLaws)(genPerson)
```

Mainly thinking this would be useful _downstream_. 😉 

The other thing is what to do when a law requires something else from the type. That could possibly be stuffed in `R`, but there may be a more elegant way.

**Comment by adamgfraser** [comment]
I was thinking the same thing. I will take this.

**Comment by jdegoes** [comment]
Here's another variation with capabilities (maybe too complicated, I am not sure):

```scala
trait Law[-Caps[_], -R, -A] { self =>
  def run(caps: Caps[A], gen: Gen[A]): ZIO[R, Nothing, TestResult] 

  def + [R1 <: R, A1 <: A](that: Law[R1, A1]): Law[R1, A1] = ???
}
object Law {
  type NoCaps[A] = Any
  ...
  def apply[A](label: String)(f: (A, A, A) => TestResult): Law[NoCaps, Any, A] = ???
  def withCaps[Caps[_], A](label: String)(f: (Caps[A], A, A, A) => TestResult)(implicit caps: Caps[A]): Law[Caps, Any, A] = ???
  ...
}
val hashConsistency = 
  Law.withCaps[Equal, Int]((equal, a1, a2) => 
    (hash(a1) === hash(a2)) ==> equal.equal(a1, a2) &&
    equal(a1, a2) ==> (hash(a1) === hash(a2))
```

**Comment by adamgfraser** [comment]
Resolved by #3111.
